apps/src/applab/applab.js

```diff
@@ -867,14 +861,52 @@ Applab.init = function(config) {
     }
   }.bind(this);
 
-  React.render(React.createElement(AppView, {
+  Applab.reactInitialProps_ = {
     assetUrl: studioApp.assetUrl,
     isEmbedView: !!config.embed,
+    isReadOnlyView: !!config.readonlyWorkspace,
     isShareView: !!config.share,
     renderCodeWorkspace: renderCodeWorkspace,
     renderVisualizationColumn: renderVisualizationColumn,
     onMount: onMount
-  }), document.getElementById(config.containerId));
+  };
+
+  Applab.reactMountPoint_ = document.getElementById(config.containerId);
+
+  Applab.render();
+};
+
+/**
+ * Cache of props, established during init, to use when re-rendering top-level
+ * view.  Eventually, it would be best to replace these with a Redux store.
+ * @type {Object}
+ */
+Applab.reactInitialProps_ = {};
+
+/**
+ * Element on which to mount the top-level React view.
+ * @type {Element}
+ * @private
+ */
+Applab.reactMountPoint_ = null;
+
+/**
+ * Trigger a top-level React render
+ */
+Applab.render = function () {
+  var nextProps = $.extend({}, Applab.reactInitialProps_, {
+    hideToggle: Applab.hideDesignModeToggle(),
+    hideViewDataButton: Applab.hideViewDataButton(),
+    startInDesignMode: Applab.startInDesignMode(),
+    initialScreen: designMode.getCurrentScreenId(),
+    screenIds: designMode.getAllScreenIds(),
+    onDesignModeButton: Applab.onDesignModeButton,
+    onCodeModeButton: Applab.onCodeModeButton,
+    onViewDataButton: Applab.onViewData,
+    onScreenChange: designMode.changeScreen,
+    onScreenCreate: designMode.createScreen
```

Comment:
Should all of these on* methods not be in reactInitialProps_ since presumably they wont be changing on rerender? Or is the goal here just to show which things we might be able to move into a redux store?